### ISSUE TITLE
OCSP server is currently too busy

### ISSUE THREAD

**Issue by kimlaurio** [difi/Oxalis-AS4#102]
Anyone else seen this?

Just posting in case others have experienced this and have perhaps reached out / know more about the details. Perhaps ppl/devs at DIFI and OpenPeppol are already aware of this, and pondering the alternatives...?

**Comment by OysteinLq** [comment]
Yep, we've been seeing on a significant number of our SMP lookups since Friday.
It looks like our very first instance of the error was on 2020-02-28 at 16:52:26.
I don't have a good estimate on what percentage of our requests fail. We've contacted OpenPeppol about it, after first trying Digdir/difi.

**Comment by senikk** [comment]
We also have these errors and its in the same service at same line as the problems with the SMP lookup explained in #99 

no.difi.oxalis.api.lang.OxalisTransmissionException: OCSP server is currently too busy.
              at no.difi.oxalis.outbound.lookup.CachedLookupService.lookup(CachedLookupService.java:73)
              at no.difi.oxalis.api.lookup.LookupService.lookup(LookupService.java:57)

**Comment by OysteinLq** [comment]
This is the response we got from OpenPEPPOL:

> 
> this issue is most likely related to DigiCerts OCSP server experiencing capacity problems. We have reached out to DigiCert to inform us about any problems on their side.
> 
> I did some manual OCSP querying and the service seems to be back online now, can you verify that you are able to send? I would suggest you inform your software vendor about the problems, maybe they could add a fallback mechanism for CRL bases revocation check when OCSP is down.

So there's that ...

**Comment by kimlaurio** [comment]
Well, what else are they going to say? :-) But the single point of failure remains.

I guess DIFI can proceed to add CRL/caching code, with some configurable TTL.

Do we need to create separate issue, or can this one count as a "request for workaround"?

**Comment by FrodeBjerkholt** [comment]
I don't know, but please ask the question at the Oxalis repo and not the Oxalis-AS4 repo.

**Comment by senikk** [comment]
Seems like this problem is already been reported at Oxalis repo aswell two hours ago at https://github.com/difi/oxalis/issues/460

**Comment by FrodeBjerkholt** [comment]
Ok - Then I close it here.

**Comment by phax** [comment]
Hi guys,
on behalf of OpenPEPPOL we've looked at this.
We came to the conclusion that this must be a pure Oxalis issue and NOT a Digicert issue.
Our assumption is, that Oxalis performs a remote query to the OCSP server for every transmission, and if many documents are sent in parallel that may lead to an overloaded OCSP server.

Here are some suggestions on what you can do to ease the pain:
* Please cache OCSP results per X509 certificate. Check results SHOULD be cached up to 6 hours
* Implement a fallback to CRL. CRL check results SHOULD be cached up to 24 hours.
* If both fail, retry internally with a certain delay between the tries
* Try sending less documents in parallel using Oxalis
